src/audio/au_portaudio.c

```diff
@@ -0,0 +1,235 @@
+  data->bufpos += num_bytes;
+  data->num_frames = remaining_frames;
+  if (remaining_frames == 0) return paComplete;
+  return paContinue;
+}
+
+
+typedef struct cst_audio_portaudiodata_struct {
+  PaStreamParameters *outputParameters;
+  double sample_rate;
+  long int bytes_per_frame;
+  PaStream *stream;
+} cst_audio_portaudiodata;
+
+
+static int audio_error_portaudio(int err) {
+  if( err == paNoError ) {
+    return err;
+  }
+  cst_errmsg("An error occured while using the portaudio stream\n");
+  cst_errmsg("Error number: %d\n", err);
+  cst_errmsg("Error message: %s\n", Pa_GetErrorText(err));
+  return err;
+}
+
+
+cst_audiodev * audio_open_portaudio(int sps, int channels, cst_audiofmt fmt) {
+  cst_audiodev *ad = cst_alloc(cst_audiodev, 1);
+  ad->sps = sps;
+  ad->real_sps = sps;
+  ad->channels = channels;
+  ad->real_channels = channels;
+  ad->fmt = fmt;
+  ad->real_fmt = fmt;
+  ad->byteswap = 0; /* not used */
+  ad->rateconv = NULL; /* not used */
+  cst_audio_portaudiodata *hdl = cst_alloc(cst_audio_portaudiodata, 1);
+  ad->platform_data = (void*) hdl;
+  hdl->outputParameters = cst_alloc(PaStreamParameters, 1);
+
+  hdl->outputParameters->device = Pa_GetDefaultOutputDevice(); /* default output device */
+  if (hdl->outputParameters->device == paNoDevice) {
+    cst_errmsg("Error: No default output device.\n");
+    audio_error_portaudio(paInvalidDevice);
+    return NULL;
+  }
+  hdl->outputParameters->channelCount = channels; /* stereo output */
+  switch(fmt) {
+    case CST_AUDIO_LINEAR16:
+      hdl->outputParameters->sampleFormat = paInt16;
+      hdl->bytes_per_frame = channels*2;
+      break;
+    case CST_AUDIO_LINEAR8:
+      hdl->outputParameters->sampleFormat = paInt8;
+      hdl->bytes_per_frame = channels*1;
+      break;
+    case CST_AUDIO_MULAW:
+      cst_errmsg("Error: MULAW not supported in portaudio.\n");
+    default:
+      audio_error_portaudio(paSampleFormatNotSupported);
+      free(ad);
+      free(hdl->outputParameters);
+      free(hdl);
+      free(ad);
+      return NULL;
+      break;
+  }
+  hdl->outputParameters->suggestedLatency = Pa_GetDeviceInfo( hdl->outputParameters->device )->defaultLowOutputLatency;
+  hdl->outputParameters->hostApiSpecificStreamInfo = NULL;
+  hdl->sample_rate = (double) sps;
+
+  return ad;
+}
+
+int audio_drain_portaudio(cst_audiodev *ad) {
+  /* audio_write_portaudio does everything */
+  return 0;
+}
+
+int audio_flush_portaudio(cst_audiodev *ad) {
+  /* audio_write_portaudio does everything */
+  return 0;
+}
+
+int audio_close_portaudio(cst_audiodev *ad) {
+  if (ad != NULL) {
+    cst_audio_portaudiodata *hdl = (cst_audio_portaudiodata*) ad->platform_data;
+    if (hdl != NULL) {
+      if (hdl->outputParameters != NULL) {
+        free(hdl->outputParameters);
+      }
+      free(ad->platform_data);
+    }
+    free(ad);
+  }
+  return 0;
+}
+
+int audio_init_portaudio() {
+    int err = Pa_Initialize();
```

Comment:
Before and after Pa_Initialize, add an error message to ignore harmless error messages from Pa_Initialize if audio works fine